frappe/www/printview.py

```diff
@@ -52,21 +52,39 @@ def get_context(context):
 		doctype=frappe.form_dict.doctype, document=frappe.form_dict.name, file_type="PDF", method="Print"
 	)
 
-	return {
-		"body": get_rendered_template(
+	is_invalid_print = False
+	print_style = None
+	try:
+		body = get_rendered_template(
 			doc,
 			print_format=print_format,
 			meta=meta,
 			trigger_print=frappe.form_dict.trigger_print,
 			no_letterhead=frappe.form_dict.no_letterhead,
 			letterhead=letterhead,
 			settings=settings,
-		),
-		"css": get_print_style(frappe.form_dict.style, print_format),
+		)
+		print_style = get_print_style(frappe.form_dict.style, print_format)
+	except frappe.exceptions.LinkExpiredError:
+		body = frappe.get_template("templates/print_formats/print_key_expired.html").render({})
+		context.http_status_code = 410
+		is_invalid_print = True
+	except frappe.exceptions.InvalidKey:
+		body = frappe.get_template("templates/print_formats/print_key_invalid.html").render({})
+		context.http_status_code = 401
+		is_invalid_print = True
```

Comment:
Not sure if this is the best way to show errors. The `is_invalid_print` variable may not be picked up by existing functions using this API. Can't we just render a separate page for error responses instead of returning what appears to be a valid response?

Reply:
I think we can keep this... there [will be a better way](https://github.com/frappe/frappe/pull/17046) to handle errors from v14. Won't be able to backport it since it is built on the [website routing refactor](https://github.com/frappe/frappe/pull/12334) which is only done for v14. 

---
> The is_invalid_print variable may not be picked up by existing functions using this API

Also, did not get the following part... which API?

Reply:
I thought about it again. This can be kept, it is only used in GUI and only affects those cases where a key is used to access the `printview`.

Reply:
> which API?

Technically, one could have overwritten the `printview.html` in `<custom app>/www`, which makes it like an API. But I think it's acceptable 👍🏼